File: 3Sum/3sum.dart
class Solution {
  List<List<int>> threeSum(List<int> nums) {
    nums.sort((a, b) => a.compareTo(b));
    final List<List<int>> result = [];
    for (var i = 0; i < nums.length; i++) {
      if (i > 0 && nums[i] == nums[i - 1]) {
        continue;
      }

      int j = i + 1;
      int k = nums.length - 1;

      while (j < k) {
        int sum = nums[i] + nums[j] + nums[k];

        if (sum == 0) {
          result.add([nums[i], nums[j], nums[k]]);
          j++;
          k--;
          while (j < k && nums[j - 1] == nums[j]) {
            j++;
          }

          while (j < k && nums[k] == nums[k + 1]) {
            k--;
          }
        } else if (sum > 0) {
          k--;
        } else {
          j++;
        }
      }
    }
    return result;
  }
}

File: 3Sum/3sum.go
package sum

import (
	"sort"
)

func threeSum(nums []int) [][]int {
	sort.Slice(nums, func(i, j int) bool { return nums[i] < nums[j] })
	result := [][]int{}

	for i := 0; i < len(nums)-2; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}

		j, k := i+1, len(nums)-1

		for j < k {
			sum := nums[i] + nums[j] + nums[k]

			if sum == 0 {
				result = append(result, []int{nums[i], nums[j], nums[k]})
				j++
				k--

				for j < k && nums[j] == nums[j-1] {
					j++
				}

				for j < k && nums[k] == nums[k+1] {
					k--
				}

			} else if sum > 0 {
				k--
			} else {
				j++
			}

		}

	}

	return result
}
